### PR TITLE
feat(dataframe): implement "DataFrame.__setitem__"

### PR DESCRIPTION
**Pull request by AChenQ**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-28v2d5y 实现 DataFrame.__setitem__ 方法，支持列操作](https://app.clickup.com/t/28v2d5y)

File: graviti/dataframe/frame.py [hunk 1]
```diff
@@ -126,7 +126,26 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Container:
         raise KeyError(key)
 
     def __setitem__(self, key: str, value: Iterable[Any]) -> None:
-        pass
+        # self.schema.to_builtin always returns record.
+        field = self.schema.to_builtin().fields[key]  # type: ignore[attr-defined]
```
Comment (linjiX) [inline graviti/dataframe/frame.py:130]: will getting schema from columns be better?
```python
self[key].schema
```

File: graviti/dataframe/frame.py [hunk 2]
```diff
@@ -126,7 +126,26 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Container:
...
+            )
+        if len(column) != len(self):
+            raise ValueError(
+                f"Length of values ({len(column)}) does not match length of "
+                f"self DataFrame ({len(self)})"
+            )
+
+        self._columns[key] = column
+        self._column_names.append(key)
+        # TODO: Need add corresponding operations.
```
Comment (linjiX) [inline graviti/dataframe/frame.py:148]: also need update schema

File: graviti/dataframe/frame.py (hunk 1 above)
Comment (linjiX) [reply graviti/dataframe/frame.py:130]: What if adding a new column rather than replace existing column?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 3]
```diff
@@ -126,7 +126,26 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Container:
         raise KeyError(key)
 
     def __setitem__(self, key: str, value: Iterable[Any]) -> None:
-        pass
+        # self.schema.to_builtin always returns record.
+        field = self.schema.to_builtin().fields[key]  # type: ignore[attr-defined]
+        if isinstance(value, Container):
+            if not value.schema.to_pyarrow().equals(field.to_pyarrow()):
+                raise TypeError("The schema is mismatched with the given value.")
```
Comment (linjiX) [inline graviti/dataframe/frame.py:133]: I think is acceptable to replace the column with another column with different schema

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 4]
```diff
@@ -125,8 +125,30 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Container:
 
         raise KeyError(key)
 
-    def __setitem__(self, key: str, value: Iterable[Any]) -> None:
-        pass
+    def __setitem__(self, key: str, value: Any) -> None:
+        # self.schema.to_builtin always returns record.
+        if isinstance(value, Container):
+            field = value.schema.copy()
```
Comment (linjiX) [inline graviti/dataframe/frame.py:131]: I think `field` should be `name` + `type`. But here the `field` is just `type`.

File: graviti/dataframe/frame.py [hunk 5]
```diff
@@ -125,8 +125,30 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Container:
...
+        else:
+            if key in self._columns:
+                field = self[key].schema
+                column = field.container._from_pyarrow(self._pylist_to_pyarrow(value, field), field)
+            else:
+                array = pa.array(value)
+                field = pt.PortexType.from_pyarrow(array.type)
+                column = field.container._from_pyarrow(array, field)
+
+        if len(column) != len(self):
```
Comment (linjiX) [inline graviti/dataframe/frame.py:142]: can we check the length earlier?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 5 above)
Comment (AChenQ) [reply graviti/dataframe/frame.py:145]: We cannot know the length of iterable value。

File: graviti/dataframe/frame.py [hunk 6]
```diff
@@ -125,8 +126,32 @@ def __getitem__(self, key: Union[str, Iterable[str]]) -> Container:
...
         raise KeyError(key)
 
-    def __setitem__(self, key: str, value: Iterable[Any]) -> None:
-        pass
+    def __setitem__(self, key: str, value: Union[Iterable[Any], _C]) -> None:
+        # self.schema.to_builtin always returns record.
+        if isinstance(value, Container):
+            schema = value.schema.copy()
+            column = value.copy()
+        else:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:134]: `elif`？

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**